Phosphor/src/edu/columbia/cs/psl/phosphor/instrumenter/SinkTaintingMV.java

```diff
@@ -25,50 +27,76 @@ public SinkTaintingMV(MethodVisitor mv, int access, String owner, String name, S
+    }
+
     @Override
     public void visitCode() {
         super.visitCode();
         // Check every arg to see if is taint tag
         Type[] args = Type.getArgumentTypes(desc);
         int idx = isStatic ? 0 : 1; // skip over the "this" argument for non-static methods
-        boolean skipNextPrimitive = false;
         for (int i = 0; i < args.length; i++) {
-            if ((args[i].getSort() == Type.OBJECT && !args[i].getDescriptor().equals(Configuration.TAINT_TAG_DESC)) || args[i].getSort() == Type.ARRAY) {
-                if ((args[i].getSort() == Type.ARRAY && (args[i].getElementType().getSort() != Type.OBJECT || args[i].getDescriptor().equals(Configuration.TAINT_TAG_ARRAYDESC))
-                        && args[i].getDimensions() == 1) || args[i].getDescriptor().startsWith("Ledu/columbia/cs/psl/phosphor/struct/Lazy")) {
-                    if (!skipNextPrimitive) {
-                        super.visitFieldInsn(GETSTATIC, Type.getInternalName(Configuration.class), "autoTainter", Type.getDescriptor(TaintSourceWrapper.class));
-                        super.visitVarInsn(ALOAD, idx);
-                        super.visitLdcInsn(owner+"."+name+desc);
-                        super.visitMethodInsn(INVOKEVIRTUAL, Type.getInternalName(TaintSourceWrapper.class), "checkTaint", "(Ljava/lang/Object;Ljava/lang/String;)V", false);
-                    }
-                    skipNextPrimitive = !skipNextPrimitive;
-                } else {
-                    super.visitFieldInsn(GETSTATIC, Type.getInternalName(Configuration.class), "autoTainter", Type.getDescriptor(TaintSourceWrapper.class));
-                    super.visitVarInsn(ALOAD, idx);
-                    super.visitLdcInsn(owner+"."+name+desc);
-                    super.visitMethodInsn(INVOKEVIRTUAL, Type.getInternalName(TaintSourceWrapper.class), "checkTaint", "(Ljava/lang/Object;Ljava/lang/String;)V", false);
-                }
-            } else if (!skipNextPrimitive) {
-                super.visitFieldInsn(GETSTATIC, Type.getInternalName(Configuration.class), "autoTainter", Type.getDescriptor(TaintSourceWrapper.class));
-                super.visitVarInsn(Configuration.TAINT_LOAD_OPCODE, idx);
-                super.visitLdcInsn(owner+"."+name+desc);
-                super.visitMethodInsn(INVOKEVIRTUAL, Type.getInternalName(TaintSourceWrapper.class), "checkTaint", "(" + Configuration.TAINT_TAG_DESC + "Ljava/lang/String;)V", false);
-                skipNextPrimitive = true;
-            } else if (skipNextPrimitive) {
-                skipNextPrimitive = false;
+            if(args[i].getDescriptor().equals(Configuration.TAINT_TAG_DESC)) {
+                // arg is a taint tag
+                callCheckTaintTag(idx, idx + args[i].getSize(), args[i+1]);
+            } else if(args[i].getSort() == Type.OBJECT) {
+                callCheckTaintObject(idx);
```

Comment:
Should there be a case for arrays, too (this would be either non-primitive arrays, or multi-dimensional primitive arrays)?